src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -9,121 +9,173 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine
     using System.Threading.Tasks;
 
     using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine;
 
-    using CrossPlatResources = Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Resources.Resources;
+    using CrossPlatResources = Resources.Resources;
 
     /// <summary>
     /// Orchestrates test session operations for the engine communicating with the client.
     /// </summary>
     public class ProxyTestSessionManager : IProxyTestSessionManager
     {
         private readonly object lockObject = new object();
-        private int parallelLevel;
-        private bool skipDefaultAdapters;
+        private readonly object proxyOperationLockObject = new object();
+        private volatile bool proxySetupFailed = false;
+        private StartTestSessionCriteria testSessionCriteria;
+        private int testhostCount;
+        private TestSessionInfo testSessionInfo;
         private Func<ProxyOperationManager> proxyCreator;
-        private Queue<Guid> availableProxyQueue;
-        private IDictionary<Guid, ProxyOperationManagerContainer> proxyMap;
+        private IList<ProxyOperationManagerContainer> proxyContainerList;
```

Comment:
The mapping that we had in place before was wrong because we didn't map source -> proxy, but rather GUID -> proxy. As such in order to introduce source information I'd probably had to add a secondary mapping that would've mapped the source(s) to the said GUID. On top of that there was the `availableProxyQueue` that's now not needed anymore because we don't dequeue elements 'per-se', in the sense that we don't use a queue structure. We just take whichever proxy we need (i.e. the one cleared to run the current source), not the first available one in the queue, so a list to hold the proxies makes more sense now.

In order to avoid this double mapping and to simplify the logic of actually dequeueing a proxy, I introduced a list of proxies where each proxy is identified by the index in this list, and then mapped the source(s) -> proxy index.